### DISCUSSION TITLE
Exceptions become a mess in logs using a Python 3.11 docker operator

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#33052] [Q&A]
We drive a lot of our scripts via the DockerOperator as we define our Python script environment separately from our Airflow environment.

Recently we upgraded our script environment to Python 3.11 and an issue has come up using the default Airflow log capturing, Python 3.11 introduced these nice arrows: https://realpython.com/python311-error-messages/#improvements-in-python-311

But they don't always stay on one line in the Airflow logs and can end up looking a bit of a mess, e.g. here is a snippet of part of an exception in one of our logs:

```
File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
return self.main(*args, **kwargs)
^^^^
^^^^^^
^^^^^^^
^^^^^^^
^^
File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1055, in main
rv = self.invoke(ctx)
^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
return ctx.invoke(self.callback, **ctx.params)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/click/core.py", line 760, in invoke
return __callback(*args, **kwargs)
^^^^^^^^^^^^
^^^^^^^^^
^^^^^^
```

Has anyone else suffered this and fixed it? Or would someone be able to give some hints on maybe how to solve this.

**Comment by notatallshaw-gts** [comment]
I'm pretty sure the problem is this code: https://github.com/apache/airflow/blob/5387c0c761fe161f28cee70a46009e1e67286679/airflow/providers/docker/operators/docker.py#L398

Treating chunks on a stream like they are log lines and stripping them of all whitespace is not going to produce nice formatting.

I am developing a patch and running it locally, if all is successful, I will look to make a PR.

**Reply by potiuk** [reply]
Yep. Very likely

**Reply by notatallshaw-gts** [reply]
I have a patch working in our environment, we now have very pretty exceptions (especially customizing the email template to put the exception in a monospace font).

If no one else jumps on this, I'll request to make an open-source contribution next week and once approved follow-up with a PR.

**Reply by potiuk** [reply]
> If no one else jumps on this, I'll request to make an open-source contribution next week and once approved follow-up with a PR.

Would be great !

**Reply by potiuk** [reply]
And feel free to **just** submit PR - no need for approval or issues, We are very lean here - if you show up with PR and good reasoning, it will get some review, approval and merge, no problem with that @notatallshaw-gts

**Reply by gabrielcnr** [reply]
We are also using latest Airflow running tasks with DockerOperator that runs Python 3.11 inside. Super keen to see the fix for this coming soon! Thanks @notatallshaw-gts

**Reply by potiuk** [reply]
I created an isssue from that https://github.com/apache/airflow/issues/33692 and marked it as "good first issue" - if you feel like it - feel free to implement it , otherwise we will have to wait for someone who picks it up and will.

**Reply by notatallshaw-gts** [reply]
I just made an internal request to open source the patch, should be approved fine I've done this a few times with Airflow already, unless anyone wants to jump on this, I should have a PR out by the end of the day.
